SSA/Projects/InstCombine/ComWrappers.lean

```diff
@@ -0,0 +1,244 @@
+import SSA.Projects.InstCombine.Base
+
+/- Wrapper around Com, Expr constructors to easily hand-write IR -/
+namespace ComWrappers
+
+local macro "offset_in_bounds" : tactic =>
+  `(tactic|
+      (
+      simp [Ctxt.snoc]
+      )
+   )
+
+def const {Γ : Ctxt _} (w : ℕ) (n : ℤ) : Expr InstCombine.Op Γ (InstCombine.Ty.bitvec w) :=
+  Expr.mk
+    (op := InstCombine.MOp.const w n)
+    (ty_eq := rfl)
+    (args := .nil)
+    (regArgs := .nil)
+
+def not {Γ : Ctxt _} (w : ℕ) (l: Nat)
+    (lp : (Ctxt.get? Γ l = some (InstCombine.MTy.bitvec (ConcreteOrMVar.concrete w)))
+      := by offset_in_bounds):
+    Expr InstCombine.Op Γ (InstCombine.Ty.bitvec w) :=
+  Expr.mk
+    (op := InstCombine.MOp.not w)
+    (ty_eq := rfl)
+    (args := .cons ⟨l, lp⟩ .nil)
+    (regArgs := .nil)
+
+def neg {Γ : Ctxt _} (w : ℕ) (l: Nat)
+    (lp : (Ctxt.get? Γ l = some (InstCombine.MTy.bitvec (ConcreteOrMVar.concrete w)))
+      := by offset_in_bounds):
+    Expr InstCombine.Op Γ (InstCombine.Ty.bitvec w) :=
```

Comment:
We might want to add all these wrappers to one of the various simp-sets, so that `simp_alive_peephole` will be able to unfold them seamlessly

Reply:
Right. I feel these simp-sets will need further work. Currently I just get a pretty messy:

```lean
InstCombine.MOp.const (ConcreteOrMVar.concrete 1) 0[[]]
InstCombine.MOp.unary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.UnaryOp.not)[[%0]]
InstCombine.MOp.unary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.UnaryOp.neg)[[%0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.and)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.or)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.xor)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.shl)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.lshr)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.ashr)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.sub)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.add)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.mul)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.sdiv)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.udiv)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.srem)[[%0,
 ,,
 %0]]
InstCombine.MOp.binary (ConcreteOrMVar.concrete 1) (InstCombine.MOp.BinaryOp.urem)[[%0,
 ,,
 %0]]
InstCombine.MOp.icmp (LLVM.IntPredicate.eq) (ConcreteOrMVar.concrete 1)[[%0,
 ,,
 %0]]
InstCombine.MOp.select (ConcreteOrMVar.concrete 1)[[%0,
 ,,
 %1,
 ,,
 %1]]
return %0
```

I would prefer to merge this now and iterate on a new PR to polish the simp sets.

Reply:
The "messy" thing you showed seems like exactly the form we'd want it to simplify to? It's a bit verbose, but when applied to `OpDenote.denote` (as it should be when we call `simp_alive_peephole`), that is exactly the form it needs to be in to reduce further into just the semantics, so the messy form shouldn't be exposed.

That said, if you feel strongly about it, there's not that much harm in requiring these wrappers to be unfolded away prior to calling `simp_alive_peephole`, so feel free to merge

Reply:
Great. This means that this is already simplified as needed. I am happy to iterate further on this, if you find ways to improve this.